akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala

```diff
@@ -36,240 +35,135 @@ private object PoolSlot {
 
-    /**
-     * How PoolProcessor changes its `receive`:
-     * waitingExposedPublisher -> waitingForSubscribePending -> unconnected ->
-     * waitingForDemandFromConnection OR waitingEagerlyConnected -> running
-     * Given slot can become get to 'running' state via 'waitingForDemandFromConnection' or 'waitingEagerlyConnected'.
-     * The difference between those two paths is that the first one is lazy - reacts to DispatchCommand and then uses
-     * inport and outport actors to obtain more items.
-     * Where the second one is eager - reacts to SlotShouldConnectCommand from PoolConductor, sends SlotEvent.ConnectedEagerly
-     * back to conductor and then waits for the first DispatchCommand
-     */
-    override def receive = waitingExposedPublisher
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler {
+      private var firstRequest: RequestContext = _
+      private val inflightRequests = new java.util.ArrayDeque[RequestContext]()
 
-    def waitingExposedPublisher: Receive = {
-      case ExposedPublisher(publisher) ⇒
-        exposedPublisher = publisher
-        context.become(waitingForSubscribePending)
-      case other ⇒ throw new IllegalStateException(s"The first message must be `ExposedPublisher` but was [$other]")
-    }
+      private var connectionFlowSource: SubSourceOutlet[HttpRequest] = _
+      private var connectionFlowSink: SubSinkInlet[HttpResponse] = _
 
-    def waitingForSubscribePending: Receive = {
-      case SubscribePending ⇒
-        exposedPublisher.takePendingSubscribers() foreach (s ⇒ self ! ActorPublisher.Internal.Subscribe(s))
-        log.debug("become unconnected, from subscriber pending")
-        context.become(unconnected)
-    }
+      private var isConnected = false
 
-    val unconnected: Receive = {
-      case OnNext(DispatchCommand(rc: RequestContext)) ⇒
-        val (connInport, connOutport) = runnableGraph.run()
-        connOutport ! Request(totalDemand)
-        context.become(waitingForDemandFromConnection(connInport = connInport, connOutport = connOutport, rc))
+      def disconnect(ex: Option[Throwable] = None) = {
+        connectionFlowSource.complete()
+        if (isConnected) {
+          isConnected = false
 
-      case OnNext(ConnectEagerlyCommand) ⇒
-        val (in, out) = runnableGraph.run()
-        onNext(SlotEvent.ConnectedEagerly(slotIx) :: Nil)
-        out ! Request(totalDemand)
-        context.become(waitingEagerlyConnected(connInport = in, connOutport = out))
+          val (inflightRetry, inflightFail) = if (ex.isEmpty) (inflightRequests.asScala, Nil)
+          else inflightRequests.iterator().asScala.partition(_.retriesLeft > 0)
 
-      case Request(_) ⇒ if (remainingRequested == 0) request(1) // ask for first request if necessary
+          val retries = inflightRetry.map(rc ⇒ SlotEvent.RetryRequest(if (rc.retriesLeft > 0) rc.copy(retriesLeft = rc.retriesLeft - 1) else rc)).toList
```

Comment:
Sorry, yes if ex.isEmpty then we should always retry the request regardless of the number of retries since the request was never actually issued.  The substream will request one, then when that is sent and a response is read it will request a second.  If that first request did a connection close then the stream just completes without an error, theres no network traffic out.

The retry logic applies if there is a connection refused (although that case is a bit silly since the retries are super fast its unlikely that there'd be any change) or if there's a network error during the request write or the response header read (a response body read will show up on the entity stream as an error unless we have a `Strict` response).

I've got a small refactor here that makes this a bit clearer and remvoes the `UnexpectedDisconnectException` since that actually can't happen.